File: editcredit/src/main/java/mostafa/ma/saleh/gmail/com/editcredit/EditCredit.java
package mostafa.ma.saleh.gmail.com.editcredit;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import androidx.annotation.IntRange;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.core.widget.TextViewCompat;
import androidx.appcompat.widget.AppCompatEditText;
import android.text.InputFilter;
import android.text.InputType;
import android.text.TextUtils;
import android.text.method.DigitsKeyListener;
import android.util.AttributeSet;
import android.util.SparseArray;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class EditCredit extends AppCompatEditText {

    public enum Separator {
        NONE, SPACES, DASHES
    }

    public enum Gravity {
        START, END, LEFT, RIGHT
    }

    public enum Card {
        VISA(1), MASTERCARD(2), AMEX(4), DISCOVER(8);
        private int value;

        Card(int value) {
            this.value = value;
        }
    }

    @Deprecated
    public static final int NO_SEPARATOR = 0;
    @Deprecated
    public static final int SPACES_SEPARATOR = 1;
    @Deprecated
    public static final int DASHES_SEPARATOR = 2;

    @Deprecated
    public static final int NONE = 0;
    @Deprecated
    public static final int VISA = 1;
    @Deprecated
    public static final int MASTERCARD = 2;
    @Deprecated
    public static final int AMEX = 4;
    @Deprecated
    public static final int DISCOVER = 8;

    private SparseArray<Pattern> mCCPatterns = null;

    @Nullable
    private String mSeparator;
    private boolean isValidCard;

    private Gravity mDrawableGravity = Gravity.END;
    private int mCurrentDrawableResId = 0;

    public EditCredit(Context context) {
        super(context);
        init();
    }

    public EditCredit(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
        applyAttributes(attrs);
    }

    public EditCredit(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
        applyAttributes(attrs);
    }

    private void init() {
        if (mCCPatterns == null) {
            setDisabledCards();
        }
        setInputType(InputType.TYPE_CLASS_PHONE);
        setSeparator(Separator.NONE);
        setDrawableGravity(Gravity.END);
    }

    private void applyAttributes(AttributeSet attrs) {
        TypedArray a = getContext().getTheme().obtainStyledAttributes(
                attrs,
                R.styleable.EditCredit,
                0, 0);
        try {
            setSeparator(Separator.values()[a.getInt(R.styleable.EditCredit_separator, Separator.NONE.ordinal())]);
            setDisabledCardsInternal(a.getInt(R.styleable.EditCredit_disabledCards, 0));
            setDrawableGravity(Gravity.values()[a.getInt(R.styleable.EditCredit_drawableGravity, Gravity.END.ordinal())]);
        } finally {
            a.recycle();
        }
    }

    @Override
    protected void onTextChanged(CharSequence text, int start, int lengthBefore, int lengthAfter) {

        if (mCCPatterns == null) {
            init();
        }

        String textWithoutSeparator = getTextWithoutSeparator();

        int mDrawableResId = 0;
        for (int i = 0; i < mCCPatterns.size(); i++) {
            int key = mCCPatterns.keyAt(i);

            Pattern p = mCCPatterns.get(key);

            Matcher m = p.matcher(textWithoutSeparator);
            if (isValidCard = m.find()) {
                mDrawableResId = key;
                break;
            }
        }
        if (mDrawableResId > 0 && mDrawableResId != mCurrentDrawableResId) {
            mCurrentDrawableResId = mDrawableResId;
        } else if (mDrawableResId == 0) {
            mCurrentDrawableResId = R.drawable.creditcard;
        }
        addDrawable();
        addSeparators();
    }

    private void addDrawable() {
        Drawable currentDrawable = ContextCompat.getDrawable(getContext(), mCurrentDrawableResId);
        if (currentDrawable != null && TextUtils.isEmpty(getError())) {
            currentDrawable = resize(currentDrawable);
            if (mDrawableGravity == null) {
                mDrawableGravity = Gravity.END;
            }
            switch (mDrawableGravity) {
                case START:
                    TextViewCompat.setCompoundDrawablesRelativeWithIntrinsicBounds(this, currentDrawable, null, null, null);
                    break;
                case RIGHT:
                    setCompoundDrawablesWithIntrinsicBounds(null, null, currentDrawable, null);
                    break;
                case LEFT:
                    setCompoundDrawablesWithIntrinsicBounds(currentDrawable, null, null, null);
                    break;
                default:
                    TextViewCompat.setCompoundDrawablesRelativeWithIntrinsicBounds(this, null, null, currentDrawable, null);
                    break;
            }
        }
    }

    private void addSeparators() {
        String text = getText().toString();
        if (mSeparator != null) {
            if (text.length() > 4 && !text.matches("(?:[0-9]{4}" + mSeparator + ")+[0-9]{1,4}")) {
                StringBuilder sp = new StringBuilder();
                int caretPosition = getSelectionEnd();
                String[] segments = splitString(text.replaceAll(mSeparator, ""));
                for (String segment : segments) {
                    sp.append(segment).append(mSeparator);
                }
                setText("");
                append(sp.delete(sp.length() - mSeparator.length(), sp.length()).toString());
                if (caretPosition < text.length())
                    setSelection(caretPosition);
            }
        }
    }

    private void removeSeparators() {
        String text = getText().toString();
        text = text.replaceAll(" ", "").replaceAll("-", "");
        setText("");
        append(text);
    }

    private String[] splitString(String s) {
        int arrayLength = (int) Math.ceil(((s.length() / (double) 4)));
        String[] result = new String[arrayLength];

        int j = 0;
        int lastIndex = result.length - 1;
        for (int i = 0; i < lastIndex; i++) {
            result[i] = s.substring(j, j + 4);
            j += 4;
        }
        result[lastIndex] = s.substring(j);

        return result;
    }

    public String getTextWithoutSeparator() {
        if (mSeparator == null) return getText().toString();
        return getText().toString().replaceAll(mSeparator, "");
    }

    /**
     * This method has been deprecated, please use {@link #setSeparator(Separator)} instead.
     */
    @Deprecated
    public void setSeparator(@IntRange(from = 0, to = 2) int separator) {
        if (separator > 2 || separator < 0)
            throw new IllegalArgumentException("The separator has to be one of the following:" +
                    "NO_SEPARATOR." +
                    "SPACES_SEPARATOR." +
                    "DASHES_SEPARATOR.");
        setSeparator(Separator.values()[separator]);
    }

    /**
     * Use this method to set the separator style.
     * The default separator is {@link Separator#NONE}.
     *
     * @param separator the style of the separator.
     */
    public void setSeparator(@NonNull Separator separator) {
        switch (separator) {
            case NONE:
                mSeparator = null;
                break;
            case SPACES:
                mSeparator = " ";
                break;
            case DASHES:
                mSeparator = "-";
                break;
        }
        if (mSeparator != null) {
            setFilters(new InputFilter[]{new InputFilter.LengthFilter(23)});
            setKeyListener(DigitsKeyListener.getInstance("0123456789" + mSeparator));
            addSeparators();
        } else {
            setFilters(new InputFilter[]{new InputFilter.LengthFilter(19)});
            setKeyListener(DigitsKeyListener.getInstance("0123456789"));
            removeSeparators();
        }
    }

    /**
     * Use this method to set the location of the card drawable.
     * The default gravity is {@link Gravity#END}.
     *
     * @param gravity the drawable location.
     */
    public void setDrawableGravity(@NonNull Gravity gravity) {
        mDrawableGravity = gravity;
        addDrawable();
    }

    private void setDisabledCardsInternal(int disabledCards) {
        List<Card> cards = new ArrayList<>();
        if (containsFlag(disabledCards, Card.VISA.value)) {
            cards.add(Card.VISA);
        }
        if (containsFlag(disabledCards, Card.MASTERCARD.value)) {
            cards.add(Card.MASTERCARD);
        }
        if (containsFlag(disabledCards, Card.AMEX.value)) {
            cards.add(Card.AMEX);
        }
        if (containsFlag(disabledCards, Card.DISCOVER.value)) {
            cards.add(Card.DISCOVER);
        }
        setDisabledCards(cards.toArray(new Card[0]));
    }

    /**
     * This method has been deprecated, please use {@link #setDisabledCards(Card...)}} instead.
     */
    @Deprecated
    public void setDisabledCards(int disabledCards) {
        setDisabledCardsInternal(disabledCards);
    }

    /**
     * Use this method to set which cards are disabled.
     * By default all supported cards are enabled.
     *
     * @param cards the cards to be disabled.
     */
    public void setDisabledCards(Card... cards) {
        mCCPatterns = new SparseArray<>();
        int disabledCards = 0;
        if (cards != null) {
            for (Card card : cards) {
                disabledCards |= card.value;
            }
        }
        if (!containsFlag(disabledCards, Card.VISA.value)) {
            mCCPatterns.put(R.drawable.visa, Pattern.compile("^4[0-9]{1,12}(?:[0-9]{6})?$"));
        }
        if (!containsFlag(disabledCards, Card.MASTERCARD.value)) {
            mCCPatterns.put(R.drawable.mastercard, Pattern.compile("^5[1-5][0-9]{0,14}$"));
        }
        if (!containsFlag(disabledCards, Card.AMEX.value)) {
            mCCPatterns.put(R.drawable.amex, Pattern.compile("^3[47][0-9]{0,13}$"));
        }
        if (!containsFlag(disabledCards, Card.DISCOVER.value)) {
            mCCPatterns.put(R.drawable.discover, Pattern.compile("^6(?:011|5[0-9]{1,2})[0-9]{0,12}$"));
        }
        onTextChanged("", 0, 0, 0);
    }

    private boolean containsFlag(int flagSet, int flag) {
        return (flagSet | flag) == flagSet;
    }

    public boolean isCardValid() {
        return getTextWithoutSeparator().length() > 12 && isValidCard;
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        boolean noDrawablesVisible = true;
        for (Drawable drawable : getCompoundDrawables()) {
            if (drawable != null) {
                noDrawablesVisible = false;
                break;
            }
        }
        if (noDrawablesVisible) {
            addDrawable();
        }
    }

    private Drawable resize(Drawable image) {
        int imageIntrinsicHeight = image.getIntrinsicHeight();
        int height = getMeasuredHeight() - (getPaddingTop() + getPaddingBottom());
        if (height <= 0) {
            return null;
        } else if (imageIntrinsicHeight > height) {
            Bitmap b = ((BitmapDrawable) image).getBitmap();
            float ratio = (float) image.getIntrinsicWidth() / (float) imageIntrinsicHeight;
            Bitmap bitmapResized = Bitmap.createScaledBitmap(b, (int) (height * ratio), height, false);
            return new BitmapDrawable(getResources(), bitmapResized);
        } else {
            return image;
        }
    }
}

File: editcredit/src/main/java/mostafa/ma/saleh/gmail/com/editcredit/EditCredit.kt
package mostafa.ma.saleh.gmail.com.editcredit

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.text.Editable
import android.text.InputFilter
import android.text.InputType
import android.text.TextWatcher
import android.text.method.DigitsKeyListener
import android.util.AttributeSet
import android.util.SparseArray
import androidx.annotation.DrawableRes
import androidx.appcompat.widget.AppCompatEditText
import androidx.core.content.ContextCompat
import androidx.core.widget.TextViewCompat
import java.util.regex.Pattern
import kotlin.properties.Delegates

class EditCredit @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = androidx.appcompat.R.attr.editTextStyle
) : AppCompatEditText(context, attrs, defStyleAttr) {

    private var mCCPatterns = SparseArray<Pattern>()
    private var isValidCard: Boolean = false
    private var mCurrentDrawableResId = Card.UNKNOWN.drawableRes
    private val maxLength = 16

    /**
     * This property sets the location of the card drawable.
     * The default gravity is [Gravity.END].
     */
    var drawableGravity: Gravity by Delegates.observable(Gravity.END) { _, _, _ ->
        addDrawable()
    }

    /**
     * This property sets the separator style.
     * The default separator is [Separator.NONE].
     */
    var separator: Separator by Delegates.observable(Separator.NONE) { _, oldValue, newValue ->
        filters =
            arrayOf<InputFilter>(InputFilter.LengthFilter(maxLength + (newValue.length * (maxLength / 4))))
        keyListener = DigitsKeyListener.getInstance("0123456789$newValue")
        val textWithoutSeparator = text.toString().replace(oldValue.toRegex(), "")
        val caretPosition = selectionStart + when (newValue == oldValue) {
            false -> (newValue.length * (selectionStart / 4))
            true -> if ((selectionStart - (selectionStart / 4)) % 4 == 0) 1 else 0
        }
        updateText(
            textWithoutSeparator.chunked(4)
                .joinToString("") { "$it$newValue" }
                .removeSuffix("$newValue")
        )
        if (caretPosition < text?.length ?: 0)
            setSelection(caretPosition)
    }

    val textWithoutSeparator
        get() = when (separator) {
            Separator.NONE -> text.toString()
            else -> text.toString().replace(separator.toRegex(), "")
        }

    val isCardValid: Boolean
        get() = textWithoutSeparator.length > 12 && isValidCard

    val cardType: Card
        get() = Card.from(mCurrentDrawableResId)

    enum class Separator(private val stringValue: String) {
        NONE(""), SPACES(" "), DASHES("-");

        override fun toString() = stringValue

        internal fun toRegex() = stringValue.toRegex()

        internal val length
            get() = stringValue.length
    }

    enum class Gravity {
        START, END, LEFT, RIGHT
    }

    enum class Card(
        internal val value: Int,
        @field:DrawableRes internal val drawableRes: Int,
        internal val regex: Regex
    ) {
        VISA(1, R.drawable.visa, Regex("^4[0-9]{1,12}(?:[0-9]{6})?$")),
        MASTERCARD(2, R.drawable.mastercard, Regex("^5[1-5][0-9]{0,14}$")),
        AMEX(4, R.drawable.amex, Regex("^3[47][0-9]{0,13}$")),
        DISCOVER(8, R.drawable.discover, Regex("^6(?:011|5[0-9]{1,2})[0-9]{0,12}$")),
        DINERS(16, R.drawable.diners, Regex("^3(?:0[0-5]|[68][0-9])[0-9]{0,11}\$")),
        UNKNOWN(-1, R.drawable.creditcard, Regex(".*"));

        companion object {
            internal fun from(@DrawableRes drawableRes: Int) =
                values().find { it.drawableRes == drawableRes } ?: UNKNOWN
        }
    }

    private val textWatcher = object : TextWatcher {
        override fun afterTextChanged(s: Editable?) {}

        override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}

        override fun onTextChanged(
            text: CharSequence,
            start: Int,
            lengthBefore: Int,
            lengthAfter: Int
        ) {
            var mDrawableResId = 0
            for (i in 0 until mCCPatterns.size()) {
                val key = mCCPatterns.keyAt(i)

                val p = mCCPatterns.get(key)

                val m = p.matcher(textWithoutSeparator)
                isValidCard = m.find()
                if (isValidCard) {
                    mDrawableResId = key
                    break
                }
            }
            if (mDrawableResId != 0 && mDrawableResId != mCurrentDrawableResId) {
                mCurrentDrawableResId = mDrawableResId
            } else if (mDrawableResId == 0) {
                mCurrentDrawableResId = Card.UNKNOWN.drawableRes
            }
            addDrawable()
            separator = separator
        }
    }

    init {
        enableAllCards()
        inputType = InputType.TYPE_CLASS_PHONE
        separator = Separator.NONE
        drawableGravity = Gravity.END
        attrs?.let { applyAttributes(it) }
        addTextChangedListener(textWatcher)
    }

    private fun applyAttributes(attrs: AttributeSet) {
        val a = context.theme.obtainStyledAttributes(
            attrs,
            R.styleable.EditCredit,
            0, 0
        )
        try {
            separator = Separator.values()[a.getInt(
                R.styleable.EditCredit_separator,
                Separator.NONE.ordinal
            )]
            setDisabledCards(
                *Card.values().filter {
                    containsFlag(
                        a.getInt(R.styleable.EditCredit_disabledCards, 0),
                        it.value
                    )
                }.toTypedArray()
            )
            drawableGravity = Gravity.values()[a.getInt(
                R.styleable.EditCredit_drawableGravity,
                Gravity.END.ordinal
            )]
        } finally {
            a.recycle()
        }
    }

    private fun addDrawable() {
        ContextCompat.getDrawable(context, mCurrentDrawableResId)
            ?.takeIf { error.isNullOrEmpty() }
            ?.resize()
            ?.let {
                when (drawableGravity) {
                    Gravity.START -> setDrawablesRelative(start = it)
                    Gravity.RIGHT -> setDrawables(right = it)
                    Gravity.LEFT -> setDrawables(left = it)
                    else -> setDrawablesRelative(end = it)
                }
            }
    }

    private fun updateText(newText: String) {
        removeTextChangedListener(textWatcher)
        setText("")
        append(newText)
        addTextChangedListener(textWatcher)
    }

    /**
     * Use this method to set which cards are disabled.
     * By default all supported cards are enabled.
     *
     * @param cards the cards to be disabled.
     */
    fun setDisabledCards(vararg cards: Card) {
        mCCPatterns.clear()
        for (card in Card.values()) {
            if (card.value == -1 || cards.contains(card)) continue
            mCCPatterns.put(card.drawableRes, Pattern.compile(card.regex.toString()))
        }
        textWatcher.onTextChanged("", 0, 0, 0)
    }

    /**
     * Use this method to enable all supported cards.
     *
     */
    fun enableAllCards() = setDisabledCards()

    private fun containsFlag(flagSet: Int, flag: Int): Boolean {
        return flagSet or flag == flagSet
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        if (compoundDrawables.all { it == null }) {
            addDrawable()
        }
    }

    private fun Drawable.resize() =
        when (val height = measuredHeight - (paddingTop + paddingBottom)) {
            in 1 until intrinsicHeight -> {
                val bitmap = (this as BitmapDrawable).bitmap
                val ratio = getIntrinsicWidth().toFloat() / intrinsicHeight.toFloat()
                val resizedBitmap =
                    Bitmap.createScaledBitmap(bitmap, (height * ratio).toInt(), height, false)
                resizedBitmap.density = Bitmap.DENSITY_NONE
                BitmapDrawable(resources, resizedBitmap)
            }
            in Int.MIN_VALUE..0 -> null
            else -> this
        }

    private fun setDrawablesRelative(
        start: Drawable? = null,
        top: Drawable? = null,
        end: Drawable? = null,
        bottom: Drawable? = null
    ) = TextViewCompat.setCompoundDrawablesRelativeWithIntrinsicBounds(
        this,
        start,
        top,
        end,
        bottom
    )

    private fun setDrawables(
        left: Drawable? = null,
        top: Drawable? = null,
        right: Drawable? = null,
        bottom: Drawable? = null
    ) = setCompoundDrawablesWithIntrinsicBounds(left, top, right, bottom)
}